app/src/main/java/de/danoeh/antennapod/activity/SingleFileAddActivity.java

```diff
@@ -0,0 +1,220 @@
+        String optionLabel = getString(R.string.singlefile_import_option);
+        intentPickAction = new Intent(Intent.ACTION_PICK);
+
+        if(!IntentUtils.isCallable(getApplicationContext(), intentPickAction)) {
+            intentPickAction.setData(null);
+            if(!IntentUtils.isCallable(getApplicationContext(), intentPickAction)) {
+                singleFileHeaderExplanation1.setVisibility(View.GONE);
+                singleFileExplanation1.setVisibility(View.GONE);
+                findViewById(R.id.divider1).setVisibility(View.GONE);
+                butChooseFilesystem.setVisibility(View.GONE);
+            }
+        }
+        if(singleFileExplanation1.getVisibility() == View.VISIBLE) {
+            singleFileHeaderExplanation1.setText(String.format(optionLabel, nextOption));
+            nextOption++;
+        }
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        StorageUtils.checkStorageAvailability(this);
+    }
+
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        super.onCreateOptionsMenu(menu);
+        return true;
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        switch (item.getItemId()) {
+            case android.R.id.home:
+                finish();
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    /*
+     * Creates an implicit intent to launch a file manager which lets
+     * the user choose a specific file to import.
+     */
+    private void chooseFileFromFilesystem() {
+        int permission = ActivityCompat.checkSelfPermission(this, android.Manifest.permission.READ_EXTERNAL_STORAGE);
+        if (permission != PackageManager.PERMISSION_GRANTED) {
+            requestPermission();
+        }
+
+        try {
+            startActivityForResult(intentPickAction, CHOOSE_SINGLE_FILE);
+        } catch (ActivityNotFoundException e) {
+            Log.e(TAG, "No activity found. Should never happen...");
+        }
+    }
+
+    /**
+      * Gets the path of the file chosen with chooseFileToImport()
+      */
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        //super.onActivityResult(requestCode, resultCode, data);
+        if (resultCode == RESULT_OK && requestCode == CHOOSE_SINGLE_FILE) {
+            Uri uri = data.getData();
+            if(uri != null && uri.toString().startsWith("/")) {
+                uri = Uri.parse("file://" + uri.toString());
+            }
+            boolean ret = importUri(uri);
+            if (ret == false) {
+                Toast.makeText(this, "Could not import", Toast.LENGTH_LONG).show();
+            }
+
+            //go back to main
+            Intent intent = new Intent(this, MainActivity.class);
+            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP
+                    | Intent.FLAG_ACTIVITY_NEW_TASK);
+            startActivity(intent);
```

Comment:
I think it is better to just call `finish()`. Otherwise this might create a strange backstack (when pressing the back button of the device). Another option would be to explicitly display a "Success" dialog.

Reply:
About calling `finish()`: sure.
About the dialog: at first I thought about a toast, but the dialog would indeed be better if I wanted more information (say, explicitly acknowledging the X of the Y files found were added to the feed).
I'd probably do this with a similar piece of code to that in the `startImport()` exception catch:
```
            new MaterialDialog.Builder(this)
                    .content(message + " " + e.getMessage())
                    .positiveText(android.R.string.ok)
                    .show();
```
Thoughts?

Reply:
My idea is to only allow to add a whole folder. That folder then creates a new "Podcast" in the sidebar and is updated automatically if you place new files in that folder. Maybe it's better to show that specific podcast after adding. Let's just keep this intent as it is and change it to the actual podcast later.

If you want to show an error message in a MaterialDialog, the intent should only be called if the error dialog is not shown (otherwise, the dialog will go away). Toast stays there for a few seconds even if the Activity changes.

Reply:
With `Intent.ACTION_OPEN_DOCUMENT_TREE`, you can pick a folder instead of a file. After that, get the `File` object of that folder and iterate over all children. Add all children the way you added one

Reply:
Aha, I was missing `ACTION_OPEN_DOCUMENT_TREE`. But let's leave that for a second commit.

Reply:
You are right, ACTION_OPEN_DOCUMENT_TREE is not a good solution. Additionally, it only works with new Android versions. I guess it's better to include a directory picker library. I found this one, which looks quite nice:
https://github.com/Angads25/android-filepicker/blob/release/Readme.md

The library can be configured to pick folders only and it returns the selected path.

Reply:
I'll try to add that, thanks.